p4-16/spec/P4-16-spec.mdk

```diff
@@ -3819,6 +3820,33 @@ table tbl {
 }
 ~ End P4Example
 
+# Function declarations { #sec-functions }
+
+Functions can be declared only at the top-level.  Functions are very
+similar to functions in languages such as C or Java.  However, P4 does
+not permit recursive functions.
+
+~ Begin P4Grammar
+functionDeclaration
+    : functionPrototype blockStatement
+    ;
+~ End P4Grammar
+
+Here is an example of a function that returns the maximum of two 32-bit values:
+
+~ Begin P4Example
+bit<32> max(in bit<32> left, in bit<32> right) {
```

Comment:
Is overloading allowed for function? Similarly, is optional argument allowed?

Reply:
yes, but that is part of the other spec PR, with the named arguments.
That will clarify the overloading rules.
And currently there are no optional arguments at all in the spec.